File: mpi_send_recv/ping_pong.c
// Author: Wes Kendall
// Copyright 2011 www.mpitutorial.com
// This code is provided freely with the tutorials on mpitutorial.com. Feel
// free to modify it for your own use. Any distribution of the code must
// either provide a link to www.mpitutorial.com or keep this header in tact.
//
// Ping pong example with MPI_Send and MPI_Recv. Two processes ping pong a
// number back and forth, incrementing it until it reaches a given value.
//
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char** argv) {
  const int PING_PONG_LIMIT = 10;

  // Initialize the MPI environment
  MPI_Init(NULL, NULL);
  // Find out rank, size
  int world_rank;
  MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
  int world_size;
  MPI_Comm_size(MPI_COMM_WORLD, &world_size);

  // We are assuming at least 2 processes for this task
  if (world_size != 2) {
    fprintf(stderr, "World size must be two for %s\n", argv[0]);
    MPI_Abort(MPI_COMM_WORLD, 1); 
  }

  int ping_pong_count = 0;
  int partner_rank = (world_rank + 1) % 2;
  while (ping_pong_count < PING_PONG_LIMIT) {
    if (world_rank == ping_pong_count % 2) {
      // Increment the ping pong count before you send it
      ping_pong_count++;
      MPI_Send(&ping_pong_count, 1, MPI_INT, partner_rank, 0, MPI_COMM_WORLD);
      printf("%d sent and incremented ping_pong_count %d to %d\n",
             world_rank, ping_pong_count, partner_rank);
    } else {
      MPI_Recv(&ping_pong_count, 1, MPI_INT, partner_rank, 0, MPI_COMM_WORLD,
               MPI_STATUS_IGNORE);
      printf("%d received ping_pong_count %d from %d\n",
             world_rank, ping_pong_count, partner_rank);
    }
  }
  MPI_Finalize();
}

File: tutorials/mpi-fortran/code/ping_pong.f90
program ping_pong
  use mpi_f08

  implicit none

  integer :: world_rank, world_size
  integer :: partner_rank
  integer :: ping_pong_count, ping_pong_limit
  type(MPI_Status) :: recv_status


  call MPI_INIT()
  call MPI_COMM_SIZE(MPI_COMM_WORLD, world_size)
  call MPI_COMM_RANK(MPI_COMM_WORLD, world_rank)

  ping_pong_count = 0
  ping_pong_limit = 10

  partner_rank = mod(world_rank + 1, 2)

  do while (ping_pong_count .lt. ping_pong_limit)
    if (world_rank .eq. mod(ping_pong_count, 2)) then
      ! Increment the ping pong count before you send it
      ping_pong_count = ping_pong_count + 1
      call MPI_SEND(ping_pong_count, 1, MPI_INT, partner_rank, 0, &
                    MPI_COMM_WORLD)
      print '(I0, " sent and incremented ping_pong_count ", I0, " to ", I0)', &
        world_rank, ping_pong_count, partner_rank
    else
      call MPI_RECV(ping_pong_count, 1, MPI_INT, partner_rank, 0, &
                    MPI_COMM_WORLD, recv_status)
      print '(I0,  " received ping_pong_count ", I0, " from ", I0)', &
        world_rank, ping_pong_count, partner_rank
    end if
  end do

  call MPI_FINALIZE()

end program
